File: web/src/lib/server/queries.ts
import {
  stages,
  type Snapshot,
  type Person,
  type HealthResult,
  type HealthState,
  type QueueItem,
  type Recommendation,
  type SearchResult,
  type Report,
} from "@/contracts/crm";
import { ApiError } from "./errors";
const day = 86400000;
const terminal = (p: Person) => ["closed", "not_interested"].includes(p.stage);
const visible = (p: Person) => !p.archivedAt && !p.mergedInto;
export function canonicalID(db: Snapshot, id: string): string {
  const people = new Map(db.people.map((p) => [p.id, p]));
  const visited = new Set<string>();
  let current = id;
  while (people.get(current)?.mergedInto) {
    if (visited.has(current)) throw new ApiError("CONFLICT");
    visited.add(current);
    current = people.get(current)!.mergedInto!;
  }
  return current;
}
export function health(
  db: Snapshot,
  personId: string,
  now = new Date(),
): HealthResult {
  const p = db.people.find((x) => x.id === personId && visible(x));
  if (!p) throw new ApiError("NOT_FOUND");
  const client = db.clients.find((x) => x.personId === p.id);
  const factors: string[] = [];
  let severity = 0;
  const age = p.lastContactAt
    ? Math.max(0, Math.floor((+now - Date.parse(p.lastContactAt)) / day))
    : null;
  factors.push(
    age === null
      ? "No meaningful contact recorded"
      : `Last meaningful contact ${age} days ago`,
  );
  if (age !== null) {
    if (age > 60) severity = 3;
    else if (age > 30) severity = 2;
  }
  const overdue = db.tasks.filter(
    (x) =>
      x.personId === p.id && x.status === "open" && Date.parse(x.dueAt) < +now,
  ).length;
  if (overdue) {
    factors.push(`${overdue} overdue commitments`);
    severity = Math.max(severity, overdue >= 3 ? 3 : 2);
  }
  if (client?.risks.length) {
    factors.push(`${client.risks.length} unresolved risks`);
    severity = 3;
  }
  if (client?.reviewAt && Date.parse(client.reviewAt) < +now) {
    factors.push("Client review is overdue");
    severity = Math.max(severity, 2);
  }
  if (
    db.onboarding.some(
      (x) =>
        x.personId === p.id &&
        x.decision === "pending" &&
        x.blockers.some((i) => !i.completed && !i.waiverReason),
    )
  ) {
    factors.push("Onboarding has unresolved blockers");
    severity = Math.max(severity, 2);
  }
  const calculated: HealthState =
    severity === 3
      ? "at_risk"
      : severity === 2
        ? "needs_attention"
        : age === null
          ? "unknown"
          : age <= 14
            ? "strong"
            : "stable";
  const state = client?.healthOverride?.state ?? calculated;
  const labels: Record<HealthState, string> = {
    at_risk: "At Risk",
    needs_attention: "Needs Attention",
    unknown: "Unknown",
    strong: "Strong",
    stable: "Stable",
  };
  return {
    personId: p.id,
    calculated,
    state,
    label: labels[state],
    explanation:
      factors.join("; ") +
      (client?.healthOverride
        ? `; CEO override: ${client.healthOverride.reason}`
        : ""),
    factors,
    evaluatedAt: now.toISOString(),
    recommendedAction: ["at_risk", "needs_attention", "unknown"].includes(state)
      ? "Review the relationship and schedule contact"
      : "Maintain the next planned review",
  };
}
export function dateKey(value: string | Date, timeZone: string): string {
  return new Intl.DateTimeFormat("en-CA", {
    timeZone,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  }).format(new Date(value));
}
export function today(
  db: Snapshot,
  now = new Date(),
  timeZone = "UTC",
): QueueItem[] {
  const items: QueueItem[] = [];
  const todayKey = dateKey(now, timeZone);
  const tasks = new Map<string, Snapshot["tasks"]>();
  for (const t of db.tasks) {
    const existing = tasks.get(t.personId) ?? [];
    existing.push(t);
    tasks.set(t.personId, existing);
  }
  const clients = new Map(db.clients.map((x) => [x.personId, x]));
  const onboardings = new Map(
    db.onboarding
      .filter((x) => x.decision === "pending")
      .map((x) => [x.personId, x]),
  );
  function add(
    p: Person,
    category: string,
    base: number,
    date: string,
    reason: string,
    action: string,
    taskId?: string,
  ) {
    const age = Math.min(
      30,
      Math.max(0, Math.floor((+now - Date.parse(date)) / day)),
    );
    const score = base + { low: 0, normal: 10, high: 20 }[p.priority] + age;
    items.push({
      id: `${p.id}:${category}:${taskId ?? "person"}`,
      personId: p.id,
      ...(taskId ? { taskId } : {}),
      category,
      score,
      reason,
      dueAt: date,
      recommendedAction: action,
      accessibleLabel: `${reason}. ${action}`,
      destination: `/people/${p.id}`,
    });
  }
  for (const p of db.people.filter((p) => visible(p) && !terminal(p))) {
    if (p.nextAction) {
      const date = dateKey(p.nextAction.dueAt, timeZone);
      if (date < todayKey)
        add(
          p,
          "overdue_follow_up",
          90,
          p.nextAction.dueAt,
          "Next action is overdue",
          `Complete or reschedule ${p.nextAction.type}`,
        );
      else if (date === todayKey)
        add(
          p,
          "due_today",
          70,
          p.nextAction.dueAt,
          "Next action is due today",
          `Prepare the scheduled ${p.nextAction.type}`,
        );
    } else if (["follow_up", "pending_signup"].includes(p.stage))
      add(
        p,
        "missing_action",
        100,
        p.stageEnteredAt,
        "Required next action is missing",
        "Schedule the next action",
      );
    const o = onboardings.get(p.id);
    if (o?.blockers.some((i) => !i.completed && !i.waiverReason))
      add(
        p,
        "onboarding_blocker",
        85,
        o.dueAt,
        "Onboarding has unresolved blockers",
        "Review the checklist",
      );
    for (const t of tasks.get(p.id) ?? [])
      if (t.status === "open" && Date.parse(t.dueAt) < +now)
        add(
          p,
          "overdue_commitment",
          95,
          t.dueAt,
          "Commitment is overdue",
          "Complete or reschedule the commitment",
          t.id,
        );
    const c = clients.get(p.id);
    if (c && p.stage === "active") {
      if (c.risks.length)
        add(
          p,
          "risk",
          110,
          p.stageEnteredAt,
          "Unresolved client risk",
          "Review recorded risks",
        );
      if (c.reviewAt && Date.parse(c.reviewAt) <= +now + 7 * day)
        add(
          p,
          "review",
          60,
          c.reviewAt,
          "Client review is upcoming or overdue",
          "Prepare a relationship review",
        );
      if (!p.lastContactAt || +now - Date.parse(p.lastContactAt) > 30 * day)
        add(
          p,
          "relationship",
          75,
          p.lastContactAt ?? p.stageEnteredAt,
          "Client needs attention",
          "Schedule meaningful contact",
        );
    }
    if (
      ["new", "follow_up", "pending_signup"].includes(p.stage) &&
      +now - Date.parse(p.stageEnteredAt) > 14 * day
    )
      add(
        p,
        "stalled",
        50,
        p.stageEnteredAt,
        "No stage movement for more than 14 days",
        "Review status and next action",
      );
    const referral = db.referrals
      .filter((r) => canonicalID(db, r.personId) === p.id)
      .sort((a, b) => a.createdAt.localeCompare(b.createdAt))[0];
    if (
      p.stage === "new" &&
      referral &&
      Date.parse(p.nextAction?.dueAt ?? referral.createdAt) <= +now
    )
      add(
        p,
        "referral",
        65,
        referral.createdAt,
        "Referral awaiting first contact",
        "Review source and contact preferences",
      );
  }
  return items.sort(
    (a, b) =>
      b.score - a.score ||
      Date.parse(a.dueAt) - Date.parse(b.dueAt) ||
      a.id.localeCompare(b.id),
  );
}
export function recommendation(
  db: Snapshot,
  id: string,
  now = new Date(),
): Recommendation {
  const p = db.people.find((x) => x.id === id && visible(x));
  if (!p) throw new ApiError("NOT_FOUND");
  const script = db.scripts
    .filter((s) => s.stage === p.stage && Date.parse(s.effectiveAt) <= +now)
    .sort((a, b) => b.version - a.version || a.id.localeCompare(b.id))[0];
  if (p.communicationRestricted)
    return {
      personId: id,
      explanation: "Communication restrictions are recorded",
      action: "Review consent before outreach",
      scriptVersionId: null,
    };
  if (terminal(p))
    return {
      personId: id,
      explanation: "This relationship is closed to active work",
      action: "Review history or explicitly reopen",
      scriptVersionId: null,
    };
  if (p.stage === "pending_signup")
    return {
      personId: id,
      explanation: "Onboarding approval requires a CEO decision",
      action: "Review required items and blockers",
      scriptVersionId: script?.id ?? null,
    };
  const h = health(db, id, now);
  return {
    personId: id,
    explanation: h.explanation,
    action: p.nextAction
      ? `Prepare the scheduled ${p.nextAction.type}`
      : h.recommendedAction,
    scriptVersionId: script?.id ?? null,
  };
}
export function search(db: Snapshot, query: string): SearchResult[] {
  const q = query.trim().normalize("NFKD").toLowerCase();
  if (!q) return [];
  if (q.length > 200) throw new ApiError("VALIDATION");
  const results: SearchResult[] = [];
  function add(
    id: string,
    kind: string,
    title: string,
    body: string,
    personId: string | null,
  ) {
    const name = title.normalize("NFKD").toLowerCase();
    if (name.includes(q) || body.normalize("NFKD").toLowerCase().includes(q))
      results.push({
        id,
        kind,
        title,
        personId,
        score:
          name === q
            ? 100
            : name.startsWith(q)
              ? 80
              : name.includes(q)
                ? 60
                : 30,
      });
  }
  for (const p of db.people.filter(visible))
    add(
      p.id,
      "person",
      p.name,
      [...p.emails, ...p.phones, ...p.tags, p.notes, p.stage].join(" "),
      p.id,
    );
  for (const o of db.organizations)
    add(o.id, "organization", o.name, o.notes, null);
  for (const t of db.tasks)
    add(t.id, "task", t.title, t.evidence ?? "", canonicalID(db, t.personId));
  for (const a of db.activities)
    add(a.id, "activity", a.summary, a.kind, canonicalID(db, a.personId));
  for (const s of db.scripts) add(s.id, "script", s.name, s.content, null);
  for (const f of db.files)
    add(f.id, "file", f.displayName, "", canonicalID(db, f.personId));
  for (const r of db.referrals)
    add(
      r.id,
      "referral",
      "Referral",
      r.sourceNote,
      canonicalID(db, r.personId),
    );
  return results
    .sort((a, b) => b.score - a.score || a.id.localeCompare(b.id))
    .slice(0, 100);
}
export function report(
  db: Snapshot,
  start: string,
  end: string,
  timeZone = "UTC",
  now = new Date(),
): Report {
  const from = Date.parse(start),
    to = Date.parse(end);
  if (!Number.isFinite(from) || !Number.isFinite(to) || from >= to)
    throw new ApiError("VALIDATION");
  const includes = (date: string) =>
    Date.parse(date) >= from && Date.parse(date) < to;
  const people = db.people.filter(visible);
  const activities = db.activities.filter((x) => includes(x.createdAt));
  const due = db.tasks.filter(
    (x) => x.status !== "cancelled" && includes(x.dueAt),
  );
  const onTime = due.filter(
    (x) =>
      x.status === "completed" &&
      x.completedAt &&
      Date.parse(x.completedAt) <= Date.parse(x.dueAt),
  );
  const referrals = db.referrals.filter((x) => includes(x.createdAt));
  const counts = Object.fromEntries(
    stages.map((stage) => [
      stage,
      people.filter((p) => p.stage === stage).length,
    ]),
  ) as Report["stages"];
  const distribution: Report["health"] = {
    strong: 0,
    stable: 0,
    needs_attention: 0,
    at_risk: 0,
    unknown: 0,
  };
  for (const p of people.filter((p) => p.stage === "active"))
    distribution[health(db, p.id, now).state]++;
  const byDay: Record<string, number> = {};
  for (const a of activities) {
    const key = dateKey(a.createdAt, timeZone);
    byDay[key] = (byDay[key] ?? 0) + 1;
  }
  const movement: Record<string, number> = {};
  for (const event of db.lifecycleHistory.filter((e) =>
    includes(e.createdAt),
  )) {
    const key = `${event.fromStage}:${event.toStage}`;
    movement[key] = (movement[key] ?? 0) + 1;
  }
  const durations = new Map<Person["stage"], number[]>();
  const eventsByPerson = new Map<string, Snapshot["lifecycleHistory"]>();
  for (const event of db.lifecycleHistory) {
    const rows = eventsByPerson.get(event.personId) ?? [];
    rows.push(event);
    eventsByPerson.set(event.personId, rows);
  }
  let activeAtStart = 0,
    retainedAtEnd = 0;
  for (const p of db.people.filter((p) => !p.mergedInto)) {
    const events = (eventsByPerson.get(p.id) ?? []).sort(
      (a, b) =>
        Date.parse(a.createdAt) - Date.parse(b.createdAt) ||
        a.id.localeCompare(b.id),
    );
    const stageAt = (time: number) => {
      if (Date.parse(p.createdAt) > time) return null;
      const earlier = events
        .filter((e) => Date.parse(e.createdAt) < time)
        .at(-1);
      if (earlier) return earlier.toStage;
      return events[0]?.fromStage ?? p.stage;
    };
    if (stageAt(from) === "active") {
      activeAtStart++;
      if (stageAt(to) === "active") retainedAtEnd++;
    }
    let cursor = Date.parse(p.createdAt),
      stage = events[0]?.fromStage ?? p.stage;
    for (const event of [
      ...events,
      { createdAt: new Date(Math.min(to, +now)).toISOString(), toStage: stage },
    ]) {
      const eventTime = Date.parse(event.createdAt);
      const overlap = Math.max(
        0,
        Math.min(eventTime, to, +now) - Math.max(cursor, from),
      );
      if (overlap > 0) {
        const list = durations.get(stage) ?? [];
        list.push(overlap / day);
        durations.set(stage, list);
      }
      cursor = eventTime;
      stage = event.toStage;
    }
  }
  const averageDaysInStage: Report["averageDaysInStage"] = {};
  for (const [stage, values] of durations)
    averageDaysInStage[stage] =
      values.reduce((sum, x) => sum + x, 0) / values.length;
  const pending = db.onboarding.filter((o) => o.decision === "pending");
  const previousStart = from - (to - from),
    previous = (date: string) =>
      Date.parse(date) >= previousStart && Date.parse(date) < from;
  const extra = {
    stageMovement: movement,
    averageDaysInStage,
    overdueFollowUps: people.filter(
      (p) =>
        p.stage === "follow_up" &&
        p.nextAction &&
        Date.parse(p.nextAction.dueAt) < +now,
    ).length,
    stalledRecords: people.filter(
      (p) =>
        ["new", "follow_up", "pending_signup"].includes(p.stage) &&
        +now - Date.parse(p.stageEnteredAt) > 14 * day,
    ).length,
    onboardingPending: pending.length,
    onboardingBlocked: pending.filter((o) =>
      o.blockers.some((i) => !i.completed && !i.waiverReason),
    ).length,
    reviewsOverdue: db.clients.filter(
      (c) =>
        c.reviewAt &&
        Date.parse(c.reviewAt) < +now &&
        people.some((p) => p.id === c.personId && p.stage === "active"),
    ).length,
    dataQualityIssues: people.filter(
      (p) =>
        (!p.emails.length && !p.phones.length) ||
        (["follow_up", "pending_signup"].includes(p.stage) && !p.nextAction),
    ).length,
    integrationWarnings: db.integrations.filter((i) => i.health === "warning")
      .length,
    retention: {
      activeAtStart,
      retainedAtEnd,
      rate: activeAtStart ? retainedAtEnd / activeAtStart : null,
    },
    comparison: {
      start: new Date(previousStart).toISOString(),
      end: start,
      calls: db.activities.filter(
        (a) => a.kind === "call" && previous(a.createdAt),
      ).length,
      newApprovals: db.lifecycleHistory.filter(
        (e) =>
          e.fromStage === "pending_signup" &&
          e.toStage === "active" &&
          previous(e.createdAt),
      ).length,
      closures: db.lifecycleHistory.filter(
        (e) => e.toStage === "closed" && previous(e.createdAt),
      ).length,
      referrals: db.referrals.filter((r) => previous(r.createdAt)).length,
    },
  };
  return {
    ...extra,
    start,
    end,
    activeClients: counts.active,
    newApprovals: db.lifecycleHistory.filter(
      (x) =>
        includes(x.createdAt) &&
        x.fromStage === "pending_signup" &&
        x.toStage === "active",
    ).length,
    closures: db.lifecycleHistory.filter(
      (x) => includes(x.createdAt) && x.toStage === "closed",
    ).length,
    calls: activities.filter((x) => x.kind === "call").length,
    referrals: referrals.length,
    convertedReferrals: referrals.filter((x) =>
      db.clients.some((c) => c.personId === canonicalID(db, x.personId)),
    ).length,
    dueTasks: due.length,
    onTimeTasks: onTime.length,
    taskCompletionRate: due.length ? onTime.length / due.length : null,
    stages: counts,
    health: distribution,
    activityByDay: byDay,
  };
}

File: legacy/apple-prototype/Sources/CRMCore/Queries.swift
import Foundation

public enum HealthState: String, Codable, CaseIterable, Sendable { case strong, stable, needsAttention, atRisk, unknown
    public var label: String { switch self { case .strong: "Strong"; case .stable: "Stable"; case .needsAttention: "Needs Attention"; case .atRisk: "At Risk"; case .unknown: "Unknown" } }
}
public struct HealthResult: Equatable, Sendable {
    public let calculated: HealthState
    public let override: HealthOverride?
    public let factors: [String]
    public let evaluatedAt: Date
    public var state: HealthState { override?.state ?? calculated }
    public var label: String { state.label }
    public var explanation: String { factors.joined(separator: "; ") + (override.map { "; CEO override: \($0.reason)" } ?? "") }
    public var symbol: String { state == .atRisk ? "exclamationmark.shield" : state == .needsAttention ? "clock.badge.exclamationmark" : "heart.text.clipboard" }
    public var suggestedAction: String { [.atRisk, .needsAttention, .unknown].contains(state) ? "Review the relationship and schedule contact" : "Maintain the next planned review" }
}
public struct HealthEngine: RelationshipHealthService {
    public init() {}
    public func evaluate(_ person: Person, in database: Database, at: Date) -> HealthResult {
        let days = person.lastContactAt.map { max(0, Int(at.timeIntervalSince($0) / 86400)) }
        let overdue = database.tasks.values.filter { database.canonicalID($0.personID) == person.id && $0.status == .open && $0.dueAt < at }.count
        var factors: [String] = []; var severity = 0
        if let days { factors.append("Last meaningful contact \(days) days ago"); if days > 60 { severity = 3 } else if days > 30 { severity = 2 } }
        else { factors.append("No meaningful contact recorded") }
        if overdue > 0 { factors.append("\(overdue) overdue commitments"); severity = max(severity, overdue >= 3 ? 3 : 2) }
        if let client = person.client {
            if !client.risks.isEmpty { factors.append("\(client.risks.count) unresolved risks"); severity = 3 }
            if let review = client.reviewAt, review < at { factors.append("Client review is overdue"); severity = max(severity, 2) }
        }
        if let onboarding = person.onboarding, onboarding.decision == .pending, onboarding.blockers.contains(where: { !$0.satisfied }) { factors.append("Onboarding has unresolved blockers"); severity = max(severity, 2) }
        let calculated: HealthState = severity == 3 ? .atRisk : severity == 2 ? .needsAttention : days == nil ? .unknown : days! <= 14 ? .strong : .stable
        return HealthResult(calculated: calculated, override: person.client?.healthOverride, factors: factors, evaluatedAt: at)
    }
}
public enum QueueCategory: String, Codable, Sendable { case overdueFollowUp, dueToday, onboardingBlocker, overdueCommitment, relationship, review, stalled, missingAction, risk, referral }
public struct QueueItem: Identifiable, Equatable, Sendable {
    public let id: String
    public let category: QueueCategory
    public let score: Int
    public let reason: String
    public let date: Date
    public let personID: UUID
    public let taskID: UUID?
    public let recommendedAction: String
    public var destination: String { "timein://person/\(personID.uuidString)" }
    public var accessibleLabel: String { "\(reason). \(recommendedAction)" }
}
public struct TodayEngine: TodayQueueService {
    public init() {}
    public func queue(in database: Database, at: Date, calendar: Calendar) -> [QueueItem] {
        var result: [QueueItem] = []
        let people = database.visiblePeople.filter { !$0.stage.isTerminal }
        let tasksByPerson = Dictionary(grouping: database.tasks.values.filter { $0.status == .open }, by: { database.canonicalID($0.personID) })
        let referralsByPerson = Dictionary(grouping: database.referrals.values, by: { database.canonicalID($0.personID) })
        func item(_ person: Person, _ category: QueueCategory, _ base: Int, _ date: Date, _ reason: String, _ action: String, taskID: UUID? = nil) -> QueueItem {
            let age = min(30, max(0, Int(at.timeIntervalSince(date) / 86400)))
            return QueueItem(id: "\(person.id):\(category.rawValue):\(taskID?.uuidString ?? "person")", category: category, score: base + person.priority.rawValue * 10 + age, reason: reason, date: date, personID: person.id, taskID: taskID, recommendedAction: action)
        }
        for person in people {
            if let next = person.nextAction {
                if next.dueAt < calendar.startOfDay(for: at) { result.append(item(person, .overdueFollowUp, 90, next.dueAt, "Next action is overdue", "Complete or reschedule \(next.type.rawValue)")) }
                else if calendar.isDate(next.dueAt, inSameDayAs: at) { result.append(item(person, .dueToday, 70, next.dueAt, "\(next.type.rawValue.capitalized) due today", "Open relationship context")) }
            } else if [.followUp, .pendingSignup].contains(person.stage) { result.append(item(person, .missingAction, 100, person.stageEnteredAt, "Required next action is missing", "Schedule the next action")) }
            if let onboarding = person.onboarding, onboarding.decision == .pending {
                if onboarding.blockers.contains(where: { !$0.satisfied }) { result.append(item(person, .onboardingBlocker, 85, onboarding.dueAt, "Onboarding has unresolved blockers", "Review the onboarding checklist")) }
            }
            for task in tasksByPerson[person.id] ?? [] where task.dueAt < at { result.append(item(person, .overdueCommitment, 95 + task.priority.rawValue * 5, task.dueAt, "Commitment is overdue", "Complete or reschedule the commitment", taskID: task.id)) }
            if let client = person.client, person.stage == .active {
                if !client.risks.isEmpty { result.append(item(person, .risk, 110, person.stageEnteredAt, "Unresolved client risk", "Review the recorded risks")) }
                if let review = client.reviewAt, review <= at.addingTimeInterval(7 * 86400) { result.append(item(person, .review, 60, review, "Client review is upcoming or overdue", "Prepare a relationship review")) }
                if person.lastContactAt.map({ at.timeIntervalSince($0) > 30 * 86400 }) ?? true { result.append(item(person, .relationship, 75, person.lastContactAt ?? person.stageEnteredAt, "Client needs relationship attention", "Schedule meaningful contact")) }
            }
            if [.new, .followUp, .pendingSignup].contains(person.stage), at.timeIntervalSince(person.stageEnteredAt) > 14 * 86400 { result.append(item(person, .stalled, 50, person.stageEnteredAt, "No stage movement for more than 14 days", "Review status and next action")) }
            if person.stage == .new, let referral = referralsByPerson[person.id]?.min(by: { $0.at < $1.at }), person.nextAction?.dueAt ?? referral.at <= at { result.append(item(person, .referral, 65, referral.at, "Referral awaiting first contact", "Review source and contact preferences")) }
        }
        return result.sorted { $0.score != $1.score ? $0.score > $1.score : $0.date != $1.date ? $0.date < $1.date : $0.id < $1.id }
    }
}
public struct Recommendation: Equatable, Sendable {
    public let explanation: String
    public let action: String
    public let scriptID: UUID?
}
public struct RecommendationEngine: Sendable {
    public init() {}
    public func recommendation(for person: Person, in db: Database, at: Date) -> Recommendation {
        let script = db.scripts.values.filter { $0.stage == person.stage && $0.effectiveAt <= at && ($0.retirementAt == nil || $0.retirementAt! > at) }.sorted { $0.version == $1.version ? $0.id.uuidString < $1.id.uuidString : $0.version > $1.version }.first
        if person.communicationRestricted { return Recommendation(explanation: "Communication restrictions are recorded", action: "Review consent before any outreach", scriptID: nil) }
        if person.stage.isTerminal { return Recommendation(explanation: "This relationship is closed to active work", action: "Review history or explicitly reopen", scriptID: nil) }
        if person.stage == .pendingSignup { return Recommendation(explanation: "Onboarding approval remains a CEO decision", action: "Review required checklist items and blockers", scriptID: script?.id) }
        let health = HealthEngine().evaluate(person, in: db, at: at)
        return Recommendation(explanation: health.explanation, action: person.nextAction.map { "Prepare the scheduled \($0.type.rawValue)" } ?? health.suggestedAction, scriptID: script?.id)
    }
}
public struct SearchResult: Identifiable, Equatable, Sendable {
    public let id: String
    public let title: String
    public let kind: String
    public let score: Int
    public let personID: UUID?
}
public struct LocalSearch: SearchService {
    public init() {}
    public func search(_ query: String, in database: Database) async throws -> [SearchResult] {
        let needle = query.trimmingCharacters(in: .whitespacesAndNewlines).folding(options: [.caseInsensitive, .diacriticInsensitive], locale: Locale(identifier: "en_US_POSIX"))
        guard !needle.isEmpty else { return [] }
        var result: [SearchResult] = []
        func append(id: UUID, title: String, text: String, kind: String, personID: UUID?) {
            let name = title.folding(options: [.caseInsensitive, .diacriticInsensitive], locale: Locale(identifier: "en_US_POSIX"))
            let body = text.folding(options: [.caseInsensitive, .diacriticInsensitive], locale: Locale(identifier: "en_US_POSIX"))
            if name.contains(needle) || body.contains(needle) { result.append(SearchResult(id: "\(kind):\(id)", title: title, kind: kind, score: name == needle ? 100 : name.hasPrefix(needle) ? 80 : name.contains(needle) ? 60 : 30, personID: personID)) }
        }
        for person in database.visiblePeople {
            try Task.checkCancellation()
            append(id: person.id, title: person.name, text: ([person.notes, person.stage.label] + person.emails + person.phones + person.tags).joined(separator: " "), kind: "Person", personID: person.id)
        }
        for org in database.organizations.values where org.metadata.archivedAt == nil { append(id: org.id, title: org.name, text: org.notes, kind: "Organization", personID: nil) }
        for task in database.tasks.values { append(id: task.id, title: task.title, text: task.evidence ?? "", kind: "Task", personID: database.canonicalID(task.personID)) }
        for activity in database.activities.values { append(id: activity.id, title: activity.summary, text: activity.kind.rawValue, kind: "Activity", personID: database.canonicalID(activity.personID)) }
        for script in database.scripts.values { append(id: script.id, title: script.name, text: script.content + " " + script.campaign, kind: "Script", personID: nil) }
        for file in database.files.values where file.archivedAt == nil { append(id: file.id, title: file.displayName, text: "", kind: "File", personID: database.canonicalID(file.personID)) }
        for referral in database.referrals.values { append(id: referral.id, title: database.people[database.canonicalID(referral.personID)]?.name ?? "Referral", text: referral.sourceNote, kind: "Referral", personID: database.canonicalID(referral.personID)) }
        try Task.checkCancellation()
        return result.sorted { $0.score == $1.score ? $0.id < $1.id : $0.score > $1.score }
    }
}
public struct OperationalIssue: Identifiable, Equatable, Sendable {
    public let id: String
    public let personID: UUID?
    public let severity: String
    public let explanation: String
    public let remediation: String
    public let detectedAt: Date
}
public struct Report: Codable, Equatable, Sendable {
    public let start: Date
    public let end: Date
    public let activeClients: Int
    public let newApprovals: Int
    public let closures: Int
    public let referrals: Int
    public let convertedReferrals: Int
    public let calls: Int
    public let dueTasks: Int
    public let onTimeTasks: Int
    public let stages: [String: Int]
    public let health: [String: Int]
    public let activityByDay: [Date: Int]
    public var followUpCompletionRate: Double? { dueTasks == 0 ? nil : Double(onTimeTasks) / Double(dueTasks) }
}
public struct ReportingEngine: Sendable {
    public init() {}
    /// Interval is half-open [start, end); counts labelled current are evaluated at `at`.
    public func report(in db: Database, start: Date, end: Date, at: Date, calendar: Calendar) throws -> Report {
        guard start < end else { throw CRMError.validation("Choose an end date after the start date.") }
        func contains(_ date: Date) -> Bool { date >= start && date < end }
        let people = db.visiblePeople
        let activities = db.activities.values.filter { contains($0.at) }
        let dueTasks = db.tasks.values.filter { $0.status != .cancelled && contains($0.dueAt) }
        let referrals = db.referrals.values.filter { contains($0.at) }
        let stages = Dictionary(grouping: people, by: { $0.stage.rawValue }).mapValues(\.count)
        let health = Dictionary(grouping: people.filter { $0.stage == .active }, by: { HealthEngine().evaluate($0, in: db, at: at).state.rawValue }).mapValues(\.count)
        return Report(start: start, end: end, activeClients: people.filter { $0.stage == .active }.count, newApprovals: activities.filter { $0.toStage == .active && $0.fromStage == .pendingSignup }.count, closures: activities.filter { $0.toStage == .closed }.count, referrals: referrals.count, convertedReferrals: referrals.filter { db.people[db.canonicalID($0.personID)]?.client != nil }.count, calls: activities.filter { $0.kind == .call }.count, dueTasks: dueTasks.count, onTimeTasks: dueTasks.filter { $0.status == .completed && ($0.completedAt ?? .distantFuture) <= $0.dueAt }.count, stages: stages, health: health, activityByDay: Dictionary(grouping: activities, by: { calendar.startOfDay(for: $0.at) }).mapValues(\.count))
    }
    public func issues(in db: Database, at: Date) -> [OperationalIssue] {
        var result: [OperationalIssue] = []
        func add(_ id: String, _ person: UUID?, _ severity: String, _ explanation: String, _ remediation: String) { result.append(OperationalIssue(id: id, personID: person, severity: severity, explanation: explanation, remediation: remediation, detectedAt: at)) }
        for p in db.visiblePeople {
            if [.followUp, .pendingSignup].contains(p.stage), p.nextAction == nil { add("action:\(p.id)", p.id, "high", "Missing required next action", "Schedule an action") }
            if p.emails.isEmpty && p.phones.isEmpty { add("contact:\(p.id)", p.id, "normal", "No contact method", "Add a permitted contact method") }
            if let o = p.onboarding, o.decision == .pending {
                if o.dueAt < at { add("onboarding:\(p.id)", p.id, "high", "Onboarding is overdue", "Review due date and next action") }
                if !o.blockers.allSatisfy(\.satisfied) { add("blocker:\(p.id)", p.id, "high", "Onboarding is blocked", "Resolve or explicitly waive blockers") }
            }
            if p.stage == .active && p.onboarding?.decision != .approved { add("lifecycle:\(p.id)", p.id, "high", "Active record lacks approval", "Review lifecycle history") }
        }
        for file in db.files.values where file.archivedAt == nil {
            if let expiry = file.expiresAt, expiry <= at.addingTimeInterval(30 * 86400) { add("file:\(file.id)", db.canonicalID(file.personID), "normal", "Agreement expires within 30 days or has expired", "Review the agreement") }
        }
        for connection in db.integrations.values where connection.health == .warning { add("integration:\(connection.id)", nil, "high", "Integration needs attention", "Review connection permissions and status") }
        return result.sorted { $0.id < $1.id }
    }
}
